lib/iris/fileformats/pp.py

```diff
@@ -1955,23 +1955,38 @@ def _convert_constraints(constraints):
     constraints = iris._constraints.list_of_constraints(constraints)
     pp_constraints = {}
     unhandled_constraints = False
+
+    def _make_func(stashobj):
+	"""
+	Provides unique name-space for each lambda function's stashobj
+	variable.
+	"""
+        return lambda stash: stash==stashobj
+
     for con in constraints:
         if isinstance(con, iris.AttributeConstraint) and \
                 list(con._attributes.keys()) == ['STASH']:
             # Convert a STASH constraint.
+            # The attribute can be a STASH object, a stashcode string,or a 
+            # callable.
             stashobj = con._attributes['STASH']
-            if not isinstance(stashobj, STASH):
-                # The attribute can be a STASH object, or a stashcode string.
-                stashobj = STASH.from_msi(stashobj)
+            if callable(stashobj):
```

Comment:
This is basically the inner loop of https://github.com/SciTools/iris/blob/master/lib/iris/_constraints.py#L446. I recommend pulling that out into the AttributeConstraint, and making use of that here.